test/AllowListStorageReader.test.ts

```diff
@@ -0,0 +1,84 @@
+import { expect } from "chai";
+import { BigNumber, Contract } from "ethers";
+import { ethers, waffle } from "hardhat";
+
+const readStorage = async (
+  base: Contract,
+  reader: Contract,
+  method: string,
+  parameters?: (string | number | BigNumber)[],
+) => {
+  const encodedCall = reader.interface.encodeFunctionData(
+    method,
+    parameters || [],
+  );
+  const resultBytes = await base.callStatic.simulateDelegatecall(
+    reader.address,
+    encodedCall,
+  );
+  return reader.interface.decodeFunctionResult(method, resultBytes)[0];
```

Comment:
This is exactly how I imagined that you need to interact with the smart contract. My only issue is with the `readStorage` `method` and `parameter` types. Specifically, you can call `readStorage("foo", 2)` or `readStorage("getSolverAt", "a string parameter when a number is expected")` and this will fail at runtime.

It's possible to limit the  types that are allowed to be passed in with something like:
```typescript
import { BigNumberish } from "ethers";

type ReaderMethods = "getSolverAt" | "numSolvers"
type ReaderParameters<M> = M extends "getSolverAt" ? [BigNumberish] : M extends "numSolvers" ? [] : never;

function readStorage<M extends ReaderMethods>(base: Contract, reader: Contract, method: M, parameter: ReaderParameters<M>) {
    // ...
}

readStorage(base, reader, "asdf", [1]) // ERR: Argument of type '"asdf"' is not assignable to parameter of type 'ReaderMethods'.
readStorage(base, reader, "getSolverAt", [0, 1, 2]) // ERR: Argument of type '[number, number, number]' is not assignable to parameter of type '[BigNumber | number | string]'.
readStorage(base, reader, "numSolvers", ["value"]) // ERR: Argument of type '[string]' is not assignable to parameter of type '[]'.

readStorage(base, reader, "getSolverAt", [1]); // OK
readStorage(base, reader, "numSolvers", []); // OK
```